### ISSUE TITLE
README around enterprise rate limiting?

### ISSUE THREAD

**Issue by FeminismIsAwesome** [mperham/sidekiq#4997]
Ruby version: 3.0.2
Rails version: 6.1.4.1
Sidekiq / Pro / Enterprise version(s): enterprise 2.2.2/pro 5.2.2/ sidekiq 6.1.1

So I was starting to try and use Enterprise Rate Limiting (https://github.com/mperham/sidekiq/wiki/Ent-Rate-Limiting), when the README suggests a pattern for the first limiter as 
```
ERP_LIMIT = Sidekiq::Limiter.concurrent('erp', 50, wait_timeout: 5, lock_timeout: 30)

def perform(...)
  ERP_LIMIT.within_limit do
    # call ERP
  end
end
```

I started doing that with the window limiter instead, and started noticing a bunch of this error:

Redis::InheritedError: Tried to use a connection from a child process without reconnecting. You need to reconnect to Redis after forking or set :inherit_socket to true.

I then noticed that further down, that rather than constants its encouraging

```
def perform(user_id)
  user_throttle = Sidekiq::Limiter.window("stripe-#{user_id}", 5, :second, wait_timeout: 5)
  user_throttle.within_limit do
    # call stripe with user's account creds
  end
end
```

Is it possible that the top part of the README is out of date? Or is there something particular to the concurrent limiter that I just need to watch out for?

Thanks for a great gem!

**Comment by mperham** [comment]
Can you show me your initializer?

**Comment by FeminismIsAwesome** [comment]
I think you mean what's in my config/initializers/sidekiq.rb?

```
require 'sidekiq/middleware/i18n'
require 'sidekiq-status'

module Sidekiq::Middleware::Whodunnit
  # Get the current locale and store it in the message
  # to be sent to Sidekiq.
  class Client
    def call(worker_class, msg, queue, redis_pool)
      msg['whodunnit'] ||= PaperTrail.request.whodunnit
      yield
    end
  end

  # Pull the msg locale out and set the current thread to use it.
  class Server
    def call(worker, msg, queue)
      PaperTrail.request.whodunnit = msg['whodunnit'] || PaperTrail.request.whodunnit
      yield
    ensure
      PaperTrail.request.whodunnit = nil
    end
  end
end

redis_opts = {url: REDIS_URL, db: REDIS_DATABASES[:sidekiq], network_timeout: 300, reconnect_attempts: 0}

Sidekiq::Enterprise.unique! unless Rails.env.test?

Sidekiq.configure_client do |config|
  config.redis = redis_opts

  config.client_middleware do |chain|
    chain.add Sidekiq::Middleware::Whodunnit::Client
    chain.add Sidekiq::Status::ClientMiddleware
  end
end

Sidekiq.configure_server do |config|
  connection = (ActiveRecord::Base.configurations[Rails.env] || Rails.application.config.database_configuration[Rails.env]).dup
  connection['reaping_frequency'] = ENV['DB_REAP_FREQ'] || 5
  connection['pool'] = Sidekiq.options[:concurrency] + 2
  ActiveRecord::Base.establish_connection(connection)

  Sidekiq.logger.info("Connection Pool size for Sidekiq Server is now: #{ActiveRecord::Base.connection.pool.instance_variable_get('@size')}")

  Sidekiq.logger.level = 0 if Rails.env == 'development'

  config.redis = redis_opts

  config.client_middleware do |chain|
    chain.add Sidekiq::Middleware::Whodunnit::Client
    chain.add Sidekiq::Status::ClientMiddleware
  end

  config.server_middleware do |chain|
    chain.add Sidekiq::Middleware::Whodunnit::Server
    chain.add Sidekiq::Status::ServerMiddleware
  end
end

Sidekiq::Extensions.enable_delay!
Sidekiq::Extensions::DelayedClass.send(:include, Sidekiq::Status::Worker)
Sidekiq.default_worker_options['retry'] = 5
```

**Comment by lmansur** [comment]
I can't help you with the actual Redis error, but regarding the Iimiter initialization, both methods (constant vs local variable) work fine. The wiki encourages you to assign the limiter to a constant to avoid instantiating it every time you use it, but if your limiter is scoped to a dynamic value (such as an account ID or API token), you'll need to initialize it before every use.

**Comment by lmansur** [comment]
As far as the actual error, I'm not 100% sure about this, but this option caught my eye:
`redis_opts = { [...], reconnect_attempts: 0}`

Does the error go away if you remove the `reconnect_attempts` option?

**Comment by mperham** [comment]
I don't really understand how you can get an InheritedError unless you are manually forking the Sidekiq process. Don't do that. Or, as the error points out, add `inherit_socket: true` to your Redis options.

**Comment by FeminismIsAwesome** [comment]
reading inherit_socket: true and us having historical reasons for turning reconnect_options to 0, I think we'll just back off using this feature for now, but I appreciate the information given here if we were to try further troubleshooting.

**Comment by marckohlbrugge** [comment]
Just wanted to say thank you for posting this!

I was getting the same error and was scratching my head what it could be. It's quite hard to Google for if you're unfamiliar with how Redis works with forking etc. When I found this thread I realized I too followed the wiki's example code of creating a constant for the limiter. But that was indeed causing these issues.

I have quite a basic Sidekiq Enterprise setup (nothing special in my initializer). So I agree the documentation should be updated to explain to add `inherit_socket: true` (and ideally include what it is), or update the code example to just avoid this issue.
